### ISSUE TITLE
fork() issue on macOS (apple silicon)

### ISSUE THREAD

**Issue by ymei** [zerotier/ZeroTierOne#2123]
I just built the latest code e299eea on macOS Ventura 13.5.2 (M1 Max).  The compilation went fine, except for some warnings like

```
ld: warning: object file (zeroidc/target/libzeroidc.a(int_util.o)) was built for newer macOS version (13.0) than being linked (11.0)
ld: warning: object file (zeroidc/target/libzeroidc.a(muldc3.o)) was built for newer macOS version (13.0) than being linked (11.0)
```

But when I start the daemon `sudo ./zerotier-one -d`, I got the following error:
```
objc[17959]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called.
objc[17959]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
```
then the process exits.

I could circumvent the safety check by
```
sudo OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES ./zerotier-one -d
```
and the rest of ZeroTier seems to run fine.

How to fix this issue?

**Comment by laduke** [comment]
Haven't seen that. Can you do a `make clean` and try again?

**Comment by ymei** [comment]
Yes, it was compiled after `make clean`.  Same issue.

**Comment by glimberg** [comment]
Can't reproduce either of these

**Comment by ymei** [comment]
My build tools/libs including `rustc 1.70.0 (90c541806 2023-05-31) (built from a source tarball)` are installed via `macports`, if this helps narrowing down the problem.

**Comment by ymei** [comment]
If I remove `--mmacosx-version-min=$(MACOS_VERSION_MIN)` and `MACOSX_DEPLOYMENT_TARGET=$(MACOS_VERSION_MIN)` in the makefile, at least the `ld` warning goes away, but the `fork()` crash remains.

**Comment by glimberg** [comment]
Sounds like it could possibly be a problem with your build environment.  We use the official Rust distribution via https://rustup.rs/.  No `macports` or `brew` involved at all in our build environments.

**Comment by laduke** [comment]
Could it be using the wrong mac SDK when building zerotier-one? That's always fun.

**Comment by ymei** [comment]
> Could it be using the wrong mac SDK when building zerotier-one? That's always fun.

```
$ ls -l /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs
total 0
drwxr-xr-x  7 root  wheel  224 Mar  9  2023 MacOSX.sdk
lrwxr-xr-x  1 root  wheel   10 Jun  1 20:41 MacOSX13.3.sdk -> MacOSX.sdk
lrwxr-xr-x  1 root  wheel   10 Jun  1 20:41 MacOSX13.sdk -> MacOSX.sdk
```

**Comment by joseph-henry** [comment]
Hello. I've also tried to replicate this issue but it works fine on my system. Upon doing a cursory glance at related issues around the internet it does appear that one of Apple's OS updates changed how `fork()` handles certain conditions and now crashes. This is not an issue due to a change in ZT. If we see more reports of this we can consider a mitigation but for now I'd suggest updating your OS. (I tested successfully on Sonoma 14.2.1).
